# ADS 509 Module 1: APIs and Web Scraping

This notebook has three parts. In the first part you will pull data from the Twitter API. In the second, you will scrape lyrics from AZLyrics.com. In the last part, you'll run code that verifies the completeness of your data pull. 

For this assignment you have chosen two musical artists who have at least 100,000 Twitter followers and 20 songs with lyrics on AZLyrics.com. In this part of the assignment we pull the some of the user information for the followers of your artist and store them in text files. 


## Important Note

This assignment requires you to have a version of Tweepy that is at least version 4. The latest version is 4.10 as I write this. Critically, this version of Tweepy is *not* on the upgrade path from Version 3, so you will not be able to simply upgrade the package if you are on Version 3. Instead you will need to explicitly install version 4, which you can do with a command like this: `pip install "tweepy>=4"`. You will also be using Version 2 of the Twitter API for this assignment. 

Run the below cell. If your version of Tweepy begins with a "4", then you should be good to go. If it begins with a "3" then run the following command, found [here](https://stackoverflow.com/questions/5226311/installing-specific-package-version-with-pip), at the command line or in a cell: `pip install -Iv tweepy==4.9`. (You may want to update that version number if Tweepy has moved on past 4.9. 

In [1]:
pip show tweepy

Name: tweepy
Note: you may need to restart the kernel to use updated packages.Version: 4.12.1
Summary: Twitter library for Python
Home-page: https://www.tweepy.org/
Author: Joshua Roesslein
Author-email: tweepy@googlegroups.com

License: MIT
Location: c:\users\steph\miniconda3\lib\site-packages
Requires: requests, oauthlib, requests-oauthlib
Required-by: 


## General Assignment Instructions

These instructions are included in every assignment, to remind you of the coding standards for the class. Feel free to delete this cell after reading it. 

One sign of mature code is conforming to a style guide. We recommend the [Google Python Style Guide](https://google.github.io/styleguide/pyguide.html). If you use a different style guide, please include a cell with a link. 

Your code should be relatively easy-to-read, sensibly commented, and clean. Writing code is a messy process, so please be sure to edit your final submission. Remove any cells that are not needed or parts of cells that contain unnecessary code. Remove inessential `import` statements and make sure that all such statements are moved into the designated cell. 

Make use of non-code cells for written commentary. These cells should be grammatical and clearly written. In some of these cells you will have questions to answer. The questions will be marked by a "Q:" and will have a corresponding "A:" spot for you. *Make sure to answer every question marked with a `Q:` for full credit.* 


# Twitter API Pull

In [1]:
# for the twitter section
import tweepy
import os
import datetime
import re
from pprint import pprint

# for the lyrics scrape section
import requests
import time
from bs4 import BeautifulSoup


In [53]:
# Use this cell for any import statements you add
import json

import twit_api_keys as keys # file with my keys saved

We need bring in our API keys. Since API keys should be kept secret, we'll keep them in a file called `api_keys.py`. This file should be stored in the directory where you store this notebook. The example file is provided for you on Blackboard. The example has API keys that are _not_ functional, so you'll need to get Twitter credentials and replace the placeholder keys. 

In [69]:
client = tweepy.Client(keys.bearer_token,wait_on_rate_limit=True) #added in my own bearer_token

client

# Testing the API

The Twitter APIs are quite rich. Let's play around with some of the features before we dive into this section of the assignment. For our testing, it's convenient to have a small data set to play with. We will seed the code with the handle of John Chandler, one of the instructors in this course. His handle is `@37chandler`. Feel free to use a different handle if you would like to look at someone else's data. 

We will write code to explore a few aspects of the API: 

1. Pull some of the followers @37chandler.
1. Explore response data, which gives us information about Twitter users. 
1. Pull the last few tweets by @37chandler.


In [70]:
handle = "37chandler"
user_obj = client.get_user(username=handle)

followers = client.get_users_followers(
    # Learn about user fields here: 
    # https://developer.twitter.com/en/docs/twitter-api/data-dictionary/object-model/user
    user_obj.data.id, user_fields=["created_at","description","location",
                                   "public_metrics"]
)



Now let's explore these a bit. We'll start by printing out names, locations, following count, and followers count for these users. 

In [71]:
num_to_print = 20

for idx, user in enumerate(followers.data) :
    following_count = user.public_metrics['following_count']
    followers_count = user.public_metrics['followers_count']
    
    print(f"{user.name} lists '{user.location}' as their location.")
    print(f" Following: {following_count}, Followers: {followers_count}.")
    print()
    
    if idx >= (num_to_print - 1) :
        break
    

John chandler lists 'Decatur, GA' as their location.
 Following: 130, Followers: 10.

Frank P Seidl lists 'Twin Cities, Minnesota USA' as their location.
 Following: 37870, Followers: 37564.

Roberta lists 'Salinas' as their location.
 Following: 1894, Followers: 181.

Anna bikes MKE lists 'mke ' as their location.
 Following: 2287, Followers: 1756.

Catherine lists 'San Angelo' as their location.
 Following: 2188, Followers: 224.

Lisa lists 'None' as their location.
 Following: 2596, Followers: 831.

Lexi lists 'None' as their location.
 Following: 432, Followers: 25.

Dave Renn lists 'None' as their location.
 Following: 98, Followers: 10.

Lionel lists 'None' as their location.
 Following: 200, Followers: 200.

Megan Randall lists 'None' as their location.
 Following: 142, Followers: 98.

Jacob Salzman lists 'None' as their location.
 Following: 563, Followers: 136.

twiter not fun lists 'None' as their location.
 Following: 218, Followers: 20.

Christian Tinsley lists 'None' as th

Let's find the person who follows this handle who has the most followers. 

In [72]:
max_followers = 0

for idx, user in enumerate(followers.data) :
    followers_count = user.public_metrics['followers_count']
    
    if followers_count > max_followers :
        max_followers = followers_count
        max_follower_user = user

        
print(max_follower_user)
print(max_follower_user.public_metrics)

SpaceConscious
{'followers_count': 37564, 'following_count': 37870, 'tweet_count': 13961, 'listed_count': 305}


Let's pull some more user fields and take a look at them. The fields can be specified in the `user_fields` argument. 

In [73]:
response = client.get_user(id=user_obj.data.id,
                          user_fields=["created_at","description","location",
                                       "entities","name","pinned_tweet_id","profile_image_url",
                                       "verified","public_metrics"])

In [74]:
for field, value in response.data.items() :
    print(f"for {field} we have {value}")

for created_at we have 2009-04-18 22:08:22+00:00
for verified we have False
for username we have 37chandler
for description we have He/Him. Data scientist, urban cyclist, educator, erstwhile frisbee player. 

¯\_(ツ)_/¯
for profile_image_url we have https://pbs.twimg.com/profile_images/2680483898/b30ae76f909352dbae5e371fb1c27454_normal.png
for location we have MN
for id we have 33029025
for name we have John Chandler
for public_metrics we have {'followers_count': 185, 'following_count': 592, 'tweet_count': 1049, 'listed_count': 3}


Now a few questions for you about the user object.

### Q: How many fields are being returned in the `response` object? 

A: 9 fields are returned -- all the user_fields specified in client.get_user() 

---

### Q: Are any of the fields within the user object non-scalar? (I.e., more complicated than a simple data type like integer, float, string, boolean, etc.) 

A: Yes, below are the Non-scalar fields  (assuming strings are scalar):

'created_at' -- date/time
'public metrics' -- Set, although within it, the data are integers.
'profile_image_url' -- url, or image (ultimately)
'description' -- this a document

The rest are simple strings or binary.

---

### Q: How many friends, followers, and tweets does this user have? 

A: This user is following 592 people (or has 592 friends), has 183 followers, and 183 tweets.


In [75]:
# displaying just those public metrics asked in last question

response.data['public_metrics']

{'followers_count': 185,
 'following_count': 592,
 'tweet_count': 1049,
 'listed_count': 3}

Although you won't need it for this assignment, individual tweets can be a rich source of text-based data. To illustrate the concepts, let's look at the last few tweets for this user. You are encouraged to explore the fields that are available about Tweets.

In [76]:

response = client.get_users_tweets(user_obj.data.id)

# By default, only the ID and text fields of each Tweet will be returned
for idx, tweet in enumerate(response.data) :
    print(tweet.id)
    print(tweet.text)
    print()
    
    if idx > 10 :
        break

1611545485029810180
Happy Dia de los Reyes to all who celebrate it. https://t.co/4G7zAuwC70

1608230093071212544
RT @year_progress: ▓▓▓▓▓▓▓▓▓▓▓▓▓▓▓ 99%

1606038920604499969
RT @CoachBalto: This video is perfect. Parents please watch. Part 1/2 https://t.co/NvcBFmyFPO

1602407567036190743
RT @LindsayMasland: I had the realization that "grades are pretend" the first time I taught (as a TA).  

I was grading something with both…

1598645130075856896
RT @marinaendicott: My new favourite lawyer’s letter, just for the sheer joy of the tone.

1598156055997222912
RT @_TanHo: Hey friends, #AdventOfCode starts TONIGHT! I've organized a friendly leaderboard every year for the #rstats (and friends) commu…

1597746144108740608
If you like biking and not getting hit by muederboxes, you should consider one of these. https://t.co/0prMLbvj3b

1597734124927995904
RT @CraigTheDev: A lot of people argue that AI art isn't theft as it isn't copying the original images but referencing them like a person.…

15

## Pulling Follower Information

In this next section of the assignment, we will pull information about the followers of your two artists. We've seen above how to pull a set of followers using `client.get_users_followers`. This function has a parameter, `max_results`, that we can use to change the number of followers that we pull. Unfortunately, we can only pull 1000 followers at a time, which means we will need to handle the _pagination_ of our results. 

The return object has the `.data` field, where the results will be found. It also has `.meta`, which we use to select the next "page" in the results using the `next_token` result. I will illustrate the ideas using our user from above. 


### Rate Limiting

Twitter limits the rates at which we can pull data, as detailed in [this guide](https://developer.twitter.com/en/docs/twitter-api/rate-limits). We can make 15 user requests per 15 minutes, meaning that we can pull $4 \cdot 15 \cdot 1000 = 60000$ users per hour. I illustrate the handling of rate limiting below, though whether or not you hit that part of the code depends on your value of `handle`.  


In the below example, I'll pull all the followers, 25 at a time. (We're using 25 to illustrate the idea; when you do this set the value to 1000.) 

In [77]:
handle_followers = [] 
pulls = 0
max_pulls = 100
next_token = None

while True :

    followers = client.get_users_followers(
        user_obj.data.id, 
        max_results=1000, # when you do this for real, set this to 1000!
        pagination_token = next_token,
        user_fields=["created_at","description","location",
                     "entities","name","pinned_tweet_id","profile_image_url",
                     "verified","public_metrics"]
    )
    pulls += 1
    
    for follower in followers.data : 
        follower_row = (follower.id,follower.name,follower.created_at,follower.description)
        handle_followers.append(follower_row)
    
    if 'next_token' in followers.meta and pulls < max_pulls :
        next_token = followers.meta['next_token']
    else : 
        break
        



## Pulling Twitter Data for Your Artists

Now let's take a look at your artists and see how long it is going to take to pull all their followers. 

In [78]:
artists = dict()

for handle in ['frankturner', 'tompetty'] : 
    user_obj = client.get_user(username=handle,user_fields=["public_metrics"])
    artists[handle] = (user_obj.data.id, 
                       handle,
                       user_obj.data.public_metrics['followers_count'])
    
for artist, data in artists.items() : 
    print(f"It would take {data[2]/(1000*15*4):.2f} hours to pull all {data[2]} followers for {artist}. ")

It would take 2.79 hours to pull all 167453 followers for frankturner. 
It would take 7.94 hours to pull all 476550 followers for tompetty. 


data[2] -- is this the id for each artist? 

Depending on what you see in the display above, you may want to limit how many followers you pull. It'd be great to get at least 200,000 per artist. 

As we pull data for each artist we will write their data to a folder called "twitter", so we will make that folder if needed.

In [79]:
# Make the "twitter" folder here. 

if not os.path.isdir("twitter") : 
    os.mkdir("twitter")

In this following cells, build on the above code to pull some of the followers and their data for your two artists. As you pull the data, write the follower ids to a file called `[artist name]_followers.txt` in the "twitter" folder. For instance, for Cher I would create a file named `cher_followers.txt`. As you pull the data, also store it in an object like a list or a data frame.

In addition to creating a file that only has follower IDs in it, you will create a file that includes user data. From the response object please extract and store the following fields: 

* screen_name	
* name	
* id	
* location	
* followers_count	
* friends_count	
* description

Store the fields with one user per row in a tab-delimited text file with the name `[artist name]_follower_data.txt`. For instance, for Cher I would create a file named `cher_follower_data.txt`. 

One note: the user's description can have tabs or returns in it, so make sure to clean those out of the description before writing them to the file. I've included some example code to do that below the stub. 

In [80]:
num_followers_to_pull = 20*1000 # using this to limit number of followers we pull

In [81]:
handles = ["frankturner","tompetty"]
whitespace_pattern = re.compile(r"\s+")
user_data = dict()
follower_data = dict()



for handle in handles :
    user_data[handle] = [] # will be a list of lists
    follower_data[handle] = [] # will be a simple list of IDs
    
# Grabs the time when we start making requests to the API
start_time = datetime.datetime.now()

In [82]:
client.get_user(username = "tompetty", user_fields = ['public_metrics']).data

<User id=12449542 name=Tom Petty username=tompetty>

In [83]:
client.get_user(username = "tompetty", user_fields = ['public_metrics']).data.public_metrics

{'followers_count': 476550,
 'following_count': 83,
 'tweet_count': 6104,
 'listed_count': 2792}

In [ ]:
# full code

handles = ["frankturner", "tompetty"]
whitespace_pattern = re.compile(r"\s+")
user_data = dict()
follower_data = dict()


for handle in handles :
    user_data[handle] = [] # will be a list of lists
    follower_data[handle] = [] # will be a simple list of IDs
    
# Grabs the time when we start making requests to the API
start_time = datetime.datetime.now()



for handle in handles:
    user_obj = client.get_user(username = handle, user_fields = ["public_metrics"])
    
    
    follower_data_dict = {"id": []}
    user_data_dict = {"screen_name": [], "name": [], "id": [], "location": [], "followers_count": [],
                 "friends_count": [], "description": []}
    
    user_data[handle] = []
    follower_data[handle] = []
    
    for response in tweepy.Paginator(client.get_users_followers, user_obj.data.id, max_results = 1000, 
                                     limit = 5,
                                 user_fields = ["username", "name", "id", "location",
                                                              "public_metrics", "description"]):
        for idx, follower in enumerate(response.data):
            follow_rows = (follower.username, follower.name, follower.id,
                                 follower.location, follower['public_metrics']['followers_count'], 
                                 follower['public_metrics']['following_count'],
                                 follower.description)
            follow_ids = follower.id
            
            
        
            
            user_data_dict["screen_name"].append(follower.username)
            user_data_dict["name"].append(follower.name)
            user_data_dict["id"].append(follow_ids)
            user_data_dict["location"].append(follower.location)
            user_data_dict["followers_count"].append(follower['public_metrics']['followers_count'])
            user_data_dict["friends_count"].append(follower['public_metrics']['following_count'])
            user_data_dict["description"].append(follower.description)
            
            follower_data_dict['id'].append(follow_ids)
            
            user_data[handle].append(user_data_dict)
            follower_data[handle].append(follower_data_dict)
            

    
            
            
            if idx >= (num_followers_to_pull - 1) :
                    break
    


    
            
# Let's see how long it took to grab all follower IDs
end_time = datetime.datetime.now()
print(end_time - start_time)          

In [107]:
follower_data['tompetty']

NameError: name 'tompetty' is not defined

In [108]:
import json
import os


for i in handles:
    # Create the output file names 
    path = os.getcwd()
    
    followers_output_file = path + f"\\twitter\\{i}_followers.txt"
    user_data_output_file = path + f"\\twitter\\{i}_follower_data.txt"
    with open(followers_output_file, 'w') as f:
        f.write(json.dumps(user_data[i]))
    with open(user_data_output_file, 'w') as f:
        f.write(json.dumps(follower_data[i]))

KeyError: 'frankturner'

In [86]:
followers_output_file

'C:\\Users\\steph\\OneDrive\\Documents\\GitHub\\ads509-api-scrape\\lyrics\\twitter\\tompetty_followers.txt'

Other notes:

response.data -- user objects

hit tab to see what you can do with object

In [87]:
tricky_description = """
    Home by Warsan Shire
    
    no one leaves home unless
    home is the mouth of a shark.
    you only run for the border
    when you see the whole city
    running as well.
    
"""
# This won't work in a tab-delimited text file.

clean_description = re.sub(r"\s+"," ",tricky_description).strip()
clean_description

'Home by Warsan Shire no one leaves home unless home is the mouth of a shark. you only run for the border when you see the whole city running as well.'

---

# Lyrics Scrape

This section asks you to pull data from the Twitter API and scrape www.AZLyrics.com. In the notebooks where you do that work you are asked to store the data in specific ways. 

In [3]:
artists = {"frankturner":"https://www.azlyrics.com/f/frankturner.html",
           "tompetty":"https://www.azlyrics.com/t/tompettyandtheheartbreakers.html"} 
# we'll use this dictionary to hold both the artist name and the link on AZlyrics



## A Note on Rate Limiting

The lyrics site, www.azlyrics.com, does not have an explicit maximum on number of requests in any one time, but in our testing it appears that too many requests in too short a time will cause the site to stop returning lyrics pages. (Entertainingly, the page that gets returned seems to only have the song title to [a Tom Jones song](https://www.azlyrics.com/lyrics/tomjones/itsnotunusual.html).) 

Whenever you call `requests.get` to retrieve a page, put a `time.sleep(5 + 10*random.random())` on the next line. This will help you not to get blocked. If you _do_ get blocked, which you can identify if the returned pages are not correct, just request a lyrics page through your browser. You'll be asked to perform a CAPTCHA and then your requests should start working again. 

## Part 1: Finding Links to Songs Lyrics

That general artist page has a list of all songs for that artist with links to the individual song pages. 

Q: Take a look at the `robots.txt` page on www.azlyrics.com. (You can read more about these pages [here](https://developers.google.com/search/docs/advanced/robots/intro).) Is the scraping we are about to do allowed or disallowed by this page? How do you know? 

A: From what I received, it said Disallow: /lyricsdb/ and Disallow: /song/ 
    - so, this makes me thing that scraping lyrics and songs is disallowed this page, but perhaps its because at this current moment, my IP is blocked from making more requests.


In [9]:
# running some code around robots.txt from the book, and ensuring we can get data

import urllib.robotparser

rp = urllib.robotparser.RobotFileParser()
rp.set_url("https://www.azlyrics.com/robots.txt")
rp.read()
rp.can_fetch("*", "https://www.azlyrics.com/sitemap.xml")

True

####  this said true, and now, it is false ^^

In [50]:

# testing out a smaller loop

urls = []

r = requests.get("https://www.azlyrics.com/t/tompettyandtheheartbreakers.html")

soup = BeautifulSoup(r.text, 'html.parser')

for div in soup.find_all("div", {"id": "listAlbum"}):
    for i in div.find_all('a', href = True):
        url = i['href']
        main = "azlyrics.com"
        url = main + url
        urls.append(url)
        
    
# urls

In [11]:
# Let's set up a dictionary of lists to hold our links
from collections import defaultdict
import random

lyrics_pages = defaultdict(list)


for artist, artist_page in artists.items() :
    # request the page and sleep
    r = requests.get(artist_page)
          
    soup = BeautifulSoup(r.text, 'html.parser')
    
    time.sleep(5 + 10*random.random())
    
    lyrics_pages[artist] = []
          


    # now extract the links to lyrics pages from this page
    for div in soup.find_all("div", {"id": "listAlbum"}):
          for i in div.find_all('a', href = True):
                urls = i['href']
                # store the links `lyrics_pages` where the key is the artist and the
                # value is a list of links. 
    
                lyrics_pages[artist].append(urls)
                


In [13]:
# lyrics_pages

Let's make sure we have enough lyrics pages to scrape. 

In [14]:
for artist, lp in lyrics_pages.items() :
    assert(len(set(lp)) > 20) 

In [15]:
# Let's see how long it's going to take to pull these lyrics 
# if we're waiting `5 + 10*random.random()` seconds 
for artist, links in lyrics_pages.items() : 
    print(f"For {artist} we have {len(links)}.")
    print(f"The full pull will take for this artist will take {round(len(links)*10/3600,2)} hours.")

For frankturner we have 234.
The full pull will take for this artist will take 0.65 hours.
For tompetty we have 216.
The full pull will take for this artist will take 0.6 hours.


## Part 2: Pulling Lyrics

Now that we have the links to our lyrics pages, let's go scrape them! Here are the steps for this part. 

1. Create an empty folder in our repo called "lyrics". 
1. Iterate over the artists in `lyrics_pages`. 
1. Create a subfolder in lyrics with the artist's name. For instance, if the artist was Cher you'd have `lyrics/cher/` in your repo.
1. Iterate over the pages. 
1. Request the page and extract the lyrics from the returned HTML file using BeautifulSoup.
1. Use the function below, `generate_filename_from_url`, to create a filename based on the lyrics page, then write the lyrics to a text file with that name. 


In [16]:
# testing out smaller loop on just one lyric page

test_link = 'https://azlyrics.com/lyrics/tompettyandtheheartbreakers/rockinaroundwithyou.html'

soup = BeautifulSoup(requests.get(test_link).content, "html.parser")

lyric = soup.select_one(".ringtone ~ div").get_text(strip = True, separator = "\n").replace("\n", " ")


title = soup.title.text # this is good for title

lyric

"Why be lonely Why be blue You got me babe, I got you And I can't stop thinkin' 'bout you How I dig rockin' around with you I was waiting You came through You knew no one else will do And I can't stop thinkin' 'bout you How I dig rockin' around with you You know I dig rockin' around You know I dig rockin' around You know I dig rockin' around"

In [22]:
def generate_filename_from_link(link) :
    
    if not link :
        return None
    
    # drop the http or https and the html
    name = link.replace("https","").replace("http","")
    name = name.replace(".html","")
    name = name.replace("azlyrics.com", "")

    name = name.replace("/lyrics/","")
    
    # Replace useless chareacters with UNDERSCORE
    name = name.replace("://","").replace(".","_").replace("/","_")
    
    # tack on .txt
    name = name + ".txt"
    
    return(name)


In [23]:
# testing function

generate_filename_from_link('/lyrics/frankturner/therealdamage.html')

'frankturner_therealdamage.txt'

In [41]:
# Make the lyrics folder here. If you'd like to practice your programming, add functionality 
# that checks to see if the folder exists. If it does, then use shutil.rmtree to remove it and create a new one.
import shutil

if os.path.isdir("lyrics") : 
    shutil.rmtree("lyrics/")

os.mkdir("lyrics")



In [45]:
# changing to lyrics directory

os.chdir('/lyrics')

FileNotFoundError: [WinError 2] The system cannot find the file specified: '/lyrics'

In [46]:
# making folders for artists

if os.path.isdir("frankturner") : 
    shutil.rmtree("frankturner/")

os.mkdir("frankturner")

if os.path.isdir("tompetty") : 
    shutil.rmtree("tompetty/")

os.mkdir("tompetty")

In [51]:
for artist in lyrics_pages:
    print(artist)

frankturner
tompetty


In [67]:
url_stub = "https://www.azlyrics.com" 


path = os.getcwd()



for artist in lyrics_pages:
    total_pages = 0 
    max_pages = 21 # pulling >20 pages; don't want to be blocked from site 
    start = time.time()
    
    while total_pages < max_pages:
        
        for link in lyrics_pages[artist]:

            full_url = url_stub + link
            
            soup = BeautifulSoup(requests.get(full_url).content, "html.parser")
            
            total_pages += 1
            
            time.sleep(5 + 10*random.random())
            
            lyric = soup.select_one(".ringtone ~ div").get_text(strip = True, separator = "\n").replace("\n"," ")
            title = soup.title.text
            
            filename = generate_filename_from_link(link)
            filename_path = path + f"\\{artist}\\" + filename
            
            # lyric data
            lyric_data = title + " \n\n " + lyric
            
            with open(filename_path, 'w') as f:
                f.write(json.dumps(lyric_data))
                
            if total_pages >= max_pages:
                break
                
    print(f"Total run time for {artist} was {round((time.time() - start)/3600,2)} hours.")
            
            
        

Total run time for frankturner was 0.06 hours.
Total run time for tompetty was 0.12 hours.


---

# Evaluation

This assignment asks you to pull data from the Twitter API and scrape www.AZLyrics.com.  After you have finished the above sections , run all the cells in this notebook. Print this to PDF and submit it, per the instructions.

In [65]:
# Simple word extractor from Peter Norvig: https://norvig.com/spell-correct.html
def words(text): 
    return re.findall(r'\w+', text.lower())

---

## Checking Twitter Data

The output from your Twitter API pull should be two files per artist, stored in files with formats like `cher_followers.txt` (a list of all follower IDs you pulled) and `cher_followers_data.txt`. These files should be in a folder named `twitter` within the repository directory. This code summarizes the information at a high level to help the instructor evaluate your work. 

In [94]:
# change directory back to api_scrape overall directory

os.chdir('../')

os.getcwd()

'C:\\Users\\steph\\OneDrive\\Documents\\GitHub\\ads509-api-scrape'

In [95]:
twitter_files = os.listdir("twitter")
twitter_files = [f for f in twitter_files if f != ".DS_Store"]
artist_handles = list(set([name.split("_")[0] for name in twitter_files]))

print(f"We see two artist handles: {artist_handles[0]} and {artist_handles[1]}.")

We see two artist handles: tompetty and frankturner.


In [97]:
for artist in artist_handles :
    follower_file = artist + "_followers.txt"
    follower_data_file = artist + "_follower_data.txt"
    
    ids = open("twitter/" + follower_file,'r').readlines()
    
    print(f"We see {len(ids)-1} in your follower file for {artist}, assuming a header row.")
    
    with open("twitter/" + follower_data_file,'r') as infile :
        
        # check the headers
        headers = infile.readline().split("\t")
        
        print(f"In the follower data file ({follower_data_file}) for {artist}, we have these columns:")
        print(" : ".join(headers))
        
        description_words = []
        locations = set()
        
        
        for idx, line in enumerate(infile.readlines()) :
            line = line.strip("\n").split("\t")
            
            try : 
                locations.add(line[3])            
                description_words.extend(words(line[6]))
            except :
                pass
    
        

        print(f"We have {idx+1} data rows for {artist} in the follower data file.")

        print(f"For {artist} we have {len(locations)} unique locations.")

        print(f"For {artist} we have {len(description_words)} words in the descriptions.")
        print("Here are the five most common words:")
        print(Counter(description_words).most_common(5))

        
        print("")
        print("-"*40)
        print("")
    

We see 0 in your follower file for tompetty, assuming a header row.
In the follower data file (tompetty_follower_data.txt) for tompetty, we have these columns:
[1592010937409544192, 1615066376544157696, 51813178, 1613623782945525763, 1615050787909996544, 1608787884047323137, 1444415982823411717, 726586874, 1609209028327899136, 1615036043396268032, 1351145155630424065, 1071842266057269249, 1568319444308443137, 1586387911544606722, 1519148369905561600, 1557849482779295744, 1553326778646777856, 53497500, 1448796905207525400, 1614257183604051971, 1614995599404703746, 1342912700876279819, 1558828928550125569, 1499054980808024065, 1259497951791058946, 1508899145288982542, 1357198069, 1585288606687629312, 1614942148545183745, 149510682, 410855789, 1614961889619988480, 1570243807915958279, 1504505943916429313, 807933766995996672, 1614643232754868224, 1445864294311104522, 1564525730943455234, 4374186743, 864963111924764672, 1609181690810830857, 54492864, 797615105563205632, 1614905440373972993,

## Checking Lyrics 

The output from your lyrics scrape should be stored in files located in this path from the directory:
`/lyrics/[Artist Name]/[filename from URL]`. This code summarizes the information at a high level to help the instructor evaluate your work. 

In [61]:
os.listdir()

['frankturner', 'tompetty']

In [68]:
artist_folders = os.listdir()
artist_folders = [f for f in artist_folders if os.path.isdir(f)]

for artist in artist_folders : 
    artist_files = os.listdir(artist)
    artist_files = [f for f in artist_files if 'txt' in f or 'csv' in f or 'tsv' in f]

    print(f"For {artist} we have {len(artist_files)} files.")

    artist_words = []

    for f_name in artist_files : 
        with open(artist + "/" + f_name) as infile : 
            artist_words.extend(words(infile.read()))

            
    print(f"For {artist} we have roughly {len(artist_words)} words, {len(set(artist_words))} are unique.")


For frankturner we have 21 files.
For frankturner we have roughly 6643 words, 1316 are unique.
For tompetty we have 21 files.
For tompetty we have roughly 4143 words, 605 are unique.


Looks like all files are in the folder they should be in!